### DISCUSSION TITLE
Bootstrapping Flux with Terraform

### DISCUSSION THREAD

**Discussion by Just-Insane** [fluxcd/flux2#3716] [Q&A]
It's possible I'm misunderstanding something, but how is one *supposed* to bootstrap Flux onto a cluster, ideally using Terraform?

I have been struggling for a few days trying to get Flux set up in a semi-stable manner on a Kubernetes cluster that is deployed via Terraform. I'd like to bootstrap the cluster via Terraform to be able to go from 0 to a fully-automated cluster. There are a few parts I've been having issues with though:

1. When boot strapping with Terraform, following [this guide](https://registry.terraform.io/providers/fluxcd/flux/latest/docs/guides/github), I am able to get Flux running and start adding apps at ./kubernetes/apps, however, I can't modify the gitRepository sorce for enabling things like SOPS secret decryption, as changes are overwritten when Terraform runs. I'm not sure if using Patches would allow to modify the GitRepository? It seems like Patches and Kustomizations are seperate?
2. When bootstrapping via [this guide](https://registry.terraform.io/providers/fluxcd/flux/latest/docs/guides/bootstrap_github_ssh), bootstrap fails and I'm not sure why, it seems like the components within the cluster are not deployed properly. My understanding is that this method is new, so there could be some bugs causing me issues.

It's probably a lack of understanding, but there doesn't seem to be a way to make the Terraform bootstrap idempotent?

**Comment by mloskot** [comment]
I'm also looking into deploying Flux with SOPS on AKS and authenticate Flux via Workload Identity to fetch SOPS key from Azure Key Vault.

Currently, I'm using the deprecated [AAD Pod Identtity](https://azure.github.io/aad-pod-identity), so I patch Flux manifests with dedicated `kustomization.yaml` in order to apply the `aadpodidbinding` with `sops-akv-decryptor` which is Azure user-assigned managed identity with `Key Vault Crypto User`  role assigned:

```yaml
# kustomization.yaml
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
- gotk-aad-sops.yaml
- gotk-git-ssh-secret.yaml
- gotk-components.yaml
- gotk-github-repository.yaml
- gotk-sync.yaml
- gotk-automation.yaml

patches:
- patch: |
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: kustomize-controller
      namespace: flux-system
    spec:
      template:
        metadata:
          labels:
            aadpodidbinding: sops-akv-decryptor  # match the AzureIdentity name
        spec:
          containers:
          - name: manager
            env:
            - name: AZURE_AUTH_METHOD
              value: msi
```

I'm looking into switching to the Workload Identity where patching will be necessary but of `ServiceAccount`-s used by Flux, in order to inject annotations like `azure.workload.identity/client-id: "<client-id-of-managed-identity>"`.

I'm also looking into switching to Terraform as the Flux installation driver,

Long story short, the patching seems doable with the Terraform provider:

https://github.com/fluxcd/terraform-provider-flux/blob/fa794b3e95994460ad87d349568664f17eb9caac/examples/customize-flux/main.tf#L23-L64

It's worth to mention, the `flux_sync` data source used in that example has been deprecated as per
https://github.com/fluxcd/terraform-provider-flux/blob/fa794b3e95994460ad87d349568664f17eb9caac/CHANGELOG.md?plain=1#L115

**Reply by mloskot** [reply]
Update: see https://github.com/fluxcd/terraform-provider-flux/issues/478 issue where 
1. in https://github.com/fluxcd/terraform-provider-flux/issues/478#issuecomment-1564539954 @stefanprodan confirms that with the provider customisations are only possible using `kustomization_override`
2. in https://github.com/fluxcd/terraform-provider-flux/issues/478#issuecomment-1564578193 @adeturner kindly shows the example of patching.

This should be applicable to both use cases on Azure, AAD Pod Identity and Workload Identity, in order to set up Flux to use SOPS with keys stored in Azure Key Vault. It should be possible to project this solution to other cluster ecosystems.
